### PR TITLE
chore: validate addons parameters before deployment

### PR DESCRIPTION
**Pull request by Lou1415926**
This PR validates the parameters in addon templates, as well as the customized parameters, before the CFN deployment. 

With the wrong addons parameter config, the CFN deployment fails nevertheless. With this PR, Copilot will error out earlier so that users don't have to wait until the deployment to know.

Note that this validation happens for both the workload addon, and the environment addon.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/addon/addons.go [hunk 1]
```diff
@@ -66,7 +66,7 @@ type parser struct {
 	ws                 WorkspaceAddonsReader
 	addonsDirPath      func() string
 	addonsFilePath     func(fName string) string
-	validateParameters func(params yaml.Node) error
+	validateParameters func(need, pass yaml.Node) error
```
Comment (iamhopaul123) [inline internal/pkg/addon/addons.go]: what do you think of renaming the param to be template and parameter or something like that? From the code it seems like `need` is the template file and `pass` is the parameter file

File: internal/pkg/addon/addons.go [hunk 2]
```diff
@@ -230,22 +233,47 @@ func (p *parser) parseParameters(fNames []string) (yaml.Node, error) {
...
+func validateParameters(neededNode yaml.Node, passedNode yaml.Node, reservedKeys []string) error {
+	passed := make(map[string]yaml.Node)
+	if err := passedNode.Decode(passed); err != nil {
 		return fmt.Errorf("decode \"Parameters\" section of the parameters file: %w", err)
 	}
-	for _, key := range reservedKeys {
-		if _, ok := content[key]; ok {
+	needed := make(map[string]yaml.Node)
+	if err := neededNode.Decode(needed); err != nil {
+		return fmt.Errorf("decode \"Parameters\" section of the template file: %w", err)
```
Comment (iamhopaul123) [inline internal/pkg/addon/addons.go:246]: Do we want to wrap the error msg differently for passed and needed?

File: internal/pkg/addon/addons.go [hunk 3]
```diff
@@ -230,22 +233,47 @@ func (p *parser) parseParameters(fNames []string) (yaml.Node, error) {
...
+	}
+	// The reserved keys should present/be absent in the template/parameters file.
```
Comment (iamhopaul123) [inline internal/pkg/addon/addons.go]:
```suggestion
	// The reserved keys should be present/absent in the template/parameters file.
```

**Review by iamhopaul123 — APPROVED** [review]
LGTM just nits

File: internal/pkg/addon/addons.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/addon/addons.go:246]: They are wrapped differently. For template, it ends with "...of the template file", whereas for paramters, it ends with "... of the parameters file". Or were you suggesting other changes?

File: internal/pkg/addon/addons.go (hunk 1 above)
Comment (paragbhingre) [reply internal/pkg/addon/addons.go]: +1

**Review by paragbhingre — APPROVED** [review]
I think naming is the only thing that confused me as Penghao mentioned. Otherwise, it looks good to me. Feel free to remove label once you decide on that.

**Pull request merged**